### ISSUE TITLE
Allow different brackets spacing behaviour on deconstructing pattern

### ISSUE THREAD

**Issue by albertogasparin** [eslint/eslint#7649]
**Tell us about your environment**

* **ESLint Version:** 3.10.0
* **Node Version:** 6.9.1
* **npm Version:** 3.10.9

**What did you do?**

`object-curly-spacing` and `array-bracket-spacing` only allow one style for both definition and deconstructing patterns.
```js
// object-curly-spacing: never
var { x } = {x:'a',y:'b'}; // invalid
// array-bracket-spacing: never
var [ z ] = ['c']; // invalid
```

**What did you expect to happen?**
As the meaning of the two patters is different, I'd like to be able to specify different spacing rules while using the deconstructing patter. Moreover, currently I cannot define a consistent spacing pattern if `object-curly-spacing` is `always` and `array-bracket-spacing` is `never` (or the opposite).
The solution could be ether a new option on the two existing rules or a dedicated rule (I'd like this solution better) so we can make deconstructing spacing consistent.

```js
// deconstructing-bracket-spacing: always
// object-curly-spacing: always
var { x } = { x: 'a', y: 'b' }; // valid
// array-bracket-spacing: never
var [ z ] = ['c']; // valid
```

```js
// deconstructing-bracket-spacing: never
// object-curly-spacing: always
var {x} = { x: 'a', y: 'b' }; // valid
// array-bracket-spacing: always
var [z] = [ 'c' ]; // valid
```

This new rule could also inherit the same options as the two existing rules have, allowing nested objects/arrays spacing definition.

**Comment by vitorbal** [comment]
HI @albertogasparin, thanks for your issue!
I'm not sure we should have a new rule for destructuring spacing, but I could get behind new options to both `object-curly-spacing` and `array-bracket-spacing` that would allow configuring different spacing for destructuring patterns.
Let's see what the team thinks. Would this be something you are willing to submit a PR for, if it gets accepted?

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
